# Analysis of the motor vehicle collisions in NYC part I

## 1. Data cleaning and preparation with early data exploration

### &nbsp;&nbsp;&nbsp;&nbsp;1.1 Dataset loading from CSV file

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('nypd-motor-vehicle-collisions.csv')

C:\Users\ewsze\AppData\Local\Temp\ipykernel_17308\2587788685.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('nypd-motor-vehicle-collisions.csv')


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1612178 entries, 0 to 1612177
Data columns (total 29 columns):
 #   Column                         Non-Null Count    Dtype  
---  ------                         --------------    -----  
 0   ACCIDENT DATE                  1612178 non-null  object 
 1   ACCIDENT TIME                  1612178 non-null  object 
 2   BOROUGH                        1127553 non-null  object 
 3   ZIP CODE                       1127376 non-null  object 
 4   LATITUDE                       1415893 non-null  float64
 5   LONGITUDE                      1415893 non-null  float64
 6   LOCATION                       1415893 non-null  object 
 7   ON STREET NAME                 1298002 non-null  object 
 8   CROSS STREET NAME              1079193 non-null  object 
 9   OFF STREET NAME                219732 non-null   object 
 10  NUMBER OF PERSONS INJURED      1612161 non-null  float64
 11  NUMBER OF PERSONS KILLED       1612145 non-null  float64
 12  NUMBER OF PEDE

In [4]:
df.isnull().sum()

ACCIDENT DATE                          0
ACCIDENT TIME                          0
BOROUGH                           484625
ZIP CODE                          484802
LATITUDE                          196285
LONGITUDE                         196285
LOCATION                          196285
ON STREET NAME                    314176
CROSS STREET NAME                 532985
OFF STREET NAME                  1392446
NUMBER OF PERSONS INJURED             17
NUMBER OF PERSONS KILLED              33
NUMBER OF PEDESTRIANS INJURED          0
NUMBER OF PEDESTRIANS KILLED           0
NUMBER OF CYCLIST INJURED              0
NUMBER OF CYCLIST KILLED               0
NUMBER OF MOTORIST INJURED             0
NUMBER OF MOTORIST KILLED              0
CONTRIBUTING FACTOR VEHICLE 1       4324
CONTRIBUTING FACTOR VEHICLE 2     216051
CONTRIBUTING FACTOR VEHICLE 3    1507924
CONTRIBUTING FACTOR VEHICLE 4    1590389
CONTRIBUTING FACTOR VEHICLE 5    1606556
COLLISION_ID                           0
VEHICLE TYPE COD

First insights:
- 1,6 milion data samples (rows);
- 29 columns;
- date and time always provided;
- column [3] ZIP CODE contains mixed types;
- memory usage: 356,7+ MB;
- 8 columns describing location (BOROUGH, ZIP CODE< LATITUDE, LONGITUDE< LOCATION< ON STREET NAME< CROSS STREET NAME< OFF STREET NAME);
- 8 columns describing the number of persons injured/killed (NUMBER OF PERSONS INJURED, NUMBER OF PERSONS KILLED, NUMBER OF PEDESTRIANS INJURED, NUMBER OF PEDESTRIANS KILLED, NUMBER OF CYCLIST INJURED, NUMBER OF CYCLIST KILLED, NUMBER OF MOTORIST INJURED, NUMBER OF MOTORIST KILLED);
- 5 columns describing the factor of the collision (CONTRIBUTING FACTOR VEHICLE 1 to 5);
- 5 columns describing the type of vehicle (VEHICLE TYPE CODE 1 to 5);
- provided all values for DATETIME, NUMBER OF PEDESTRIANS/CYCLIST/MOTORIST INJURED/KILLED (some values are missing in NUMBER OF PERSONS INJURED, NUMBER OF PERSONS KILLED, propably a total number of injured/killed), COLLISION_ID;
- for higher-numbered CONTRIBUTING FACTOR VEHICLE (1-5) the number of missing values increases, the same trend is observed for VEHICLE TYPE CODE (1-5);

Settings to display all columns:

In [5]:
pd.set_option('display.max_columns', None)
df.head()

ACCIDENT DATE ACCIDENT TIME    BOROUGH ZIP CODE   LATITUDE  \
0  2019-08-05T00:00:00.000         16:30     QUEENS  11434.0  40.676052   
1  2019-08-27T00:00:00.000         16:02   BROOKLYN  11225.0  40.657780   
2  2019-08-15T00:00:00.000         17:57  MANHATTAN  10002.0  40.718143   
3  2019-08-30T00:00:00.000         21:53      BRONX  10460.0  40.840534   
4  2019-08-06T00:00:00.000          9:45  MANHATTAN  10016.0  40.745440   

   LONGITUDE                                           LOCATION  \
0 -73.790184  {'type': 'Point', 'coordinates': [-73.790184, ...   
1 -73.951096  {'type': 'Point', 'coordinates': [-73.951096, ...   
2 -73.993835  {'type': 'Point', 'coordinates': [-73.993835, ...   
3 -73.866610  {'type': 'Point', 'coordinates': [-73.86661, 4...   
4 -73.975400  {'type': 'Point', 'coordinates': [-73.9754, 40...   

                     ON STREET NAME CROSS STREET NAME  \
0                               NaN               NaN   
1                               NaN               NaN   
2  CHRYSTIE STREET                       GRAND STREET   
3                               NaN               NaN   
4  EAST 35 STREET                            2 AVENUE   

                            OFF STREET NAME  NUMBER OF PERSONS INJURED  \
0  150-08    123 AVENUE                                            0.0   
1  288       HAWTHORNE STREET                                      0.0   
2                                       NaN                        1.0   
3  1837      EAST TREMONT AVENUE                                   0.0   
4                                       NaN                        1.0   

   NUMBER OF PERSONS KILLED  NUMBER OF PEDESTRIANS INJURED  \
0                       0.0                              0   
1                       0.0                              0   
2                       0.0                              0   
3                       0.0                              0   
4                       0.0                              0   

   NUMBER OF PEDESTRIANS KILLED  NUMBER OF CYCLIST INJURED  \
0                             0                          0   
1                             0                          0   
2                             0                          0   
3                             0                          0   
4                             0                          1   

   NUMBER OF CYCLIST KILLED  NUMBER OF MOTORIST INJURED  \
0                         0                           0   
1                         0                           0   
2                         0                           1   
3                         0                           0   
4                         0                           0   

   NUMBER OF MOTORIST KILLED   CONTRIBUTING FACTOR VEHICLE 1  \
0                          0             Passing Too Closely   
1                          0             Passing Too Closely   
2                          0  Driver Inattention/Distraction   
3                          0                     Unspecified   
4                          0  Driver Inattention/Distraction   

    CONTRIBUTING FACTOR VEHICLE 2 CONTRIBUTING FACTOR VEHICLE 3  \
0                     Unspecified                           NaN   
1                     Unspecified                           NaN   
2                             NaN                           NaN   
3                     Unspecified                           NaN   
4  Driver Inattention/Distraction                           NaN   

  CONTRIBUTING FACTOR VEHICLE 4 CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID  \
0                           NaN                           NaN       4184637   
1                           NaN                           NaN       4195773   
2                           NaN                           NaN       4202457   
3                           NaN                           NaN       4198749   
4                           NaN                           NaN       4183798   

       

Specifying expected data types:

In [6]:
dtypes = {
    'BOROUGH': 'category',
    'ZIP CODE': 'object',
    'LATITUDE': 'float32',
    'LONGITUDE': 'float32',
    'LOCATION': 'category',
    'ON STREET NAME': 'object',
    'CROSS STREET NAME': 'object',
    'OFF STREET NAME': 'object',
    'NUMBER OF PERSONS INJURED': 'float32',    # some NA values
    'NUMBER OF PERSONS KILLED': 'float32',    # some NA values
    'NUMBER OF PEDESTRIANS INJURED': 'int8',
    'NUMBER OF PEDESTRIANS KILLED': 'int8',
    'NUMBER OF CYCLIST INJURED': 'int8',
    'NUMBER OF CYCLIST KILLED': 'int8',
    'NUMBER OF MOTORIST INJURED': 'int8',
    'NUMBER OF MOTORIST KILLED': 'int8',
    'CONTRIBUTING FACTOR VEHICLE 1': 'object',
    'CONTRIBUTING FACTOR VEHICLE 2': 'object',
    'CONTRIBUTING FACTOR VEHICLE 3': 'object',
    'CONTRIBUTING FACTOR VEHICLE 4': 'object',
    'CONTRIBUTING FACTOR VEHICLE 5': 'object',    
    'COLLISION_ID': 'int32',
    'VEHICLE TYPE CODE 1': 'object',
    'VEHICLE TYPE CODE 2': 'object',
    'VEHICLE TYPE CODE 3': 'object',
    'VEHICLE TYPE CODE 4': 'object',
    'VEHICLE TYPE CODE 5': 'object',
}

Reading the CSV file with provided data types:

In [7]:
df = pd.read_csv('nypd-motor-vehicle-collisions.csv', dtype=dtypes, parse_dates=['ACCIDENT DATE'])

### &nbsp;&nbsp;&nbsp;&nbsp;1.2 Handling duplicated and missing data

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.1 Creating one column DATETIME (merging ACCIDENT DATE and ACCIDENT TIME) and moving it for [0] index:

In [8]:
df['DATETIME'] = pd.to_datetime(df['ACCIDENT DATE'].astype(str) + ' ' + df['ACCIDENT TIME'])
df.drop(columns=['ACCIDENT DATE', 'ACCIDENT TIME'], inplace=True)

cols = df.columns.tolist()
cols.remove('DATETIME')
cols.insert(0, 'DATETIME')
df = df[cols]

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.2 Droping duplicates

Counting duplicates:

In [9]:
duplicate_count = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_count}")

duplicate_percentage = (duplicate_count / len(df)) * 100
print(f"Duplicates make up {duplicate_percentage:.2f}% of the original dataset.")

Number of duplicate rows: 394221
Duplicates make up 24.45% of the original dataset.


24.45% of the original dataset consists of duplicate rows.

Display the head and tail of the DataFrame to inspect its structure, and check whether any duplicates are actual duplicates:

In [10]:
df[df.duplicated()].head(2)

DATETIME   BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
50000 2019-11-20 15:40:00    QUEENS    11101  40.744129 -73.927055   
50001 2019-11-12 07:05:00  BROOKLYN    11208  40.675938 -73.868233   

                                                LOCATION  \
50000  {'type': 'Point', 'coordinates': [-73.927055, ...   
50001  {'type': 'Point', 'coordinates': [-73.86823, 4...   

                         ON STREET NAME CROSS STREET NAME OFF STREET NAME  \
50000  QUEENS BOULEVARD                         38 STREET             NaN   
50001  PITKIN AVENUE                        AUTUMN AVENUE             NaN   

       NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
50000                        0.0                       0.0   
50001                        0.0                       0.0   

       NUMBER OF PEDESTRIANS INJURED  NUMBER OF PEDESTRIANS KILLED  \
50000                              0                             0   
50001                              0                             0   

       NUMBER OF CYCLIST INJURED  NUMBER OF CYCLIST KILLED  \
50000                          0                         0   
50001                          0                         0   

       NUMBER OF MOTORIST INJURED  NUMBER OF MOTORIST KILLED  \
50000                           0                          0   
50001                           0                          0   

        CONTRIBUTING FACTOR VEHICLE 1   CONTRIBUTING FACTOR VEHICLE 2  \
50000  Driver Inattention/Distraction  Driver Inattention/Distraction   
50001                     Unspecified                     Unspecified   

      CONTRIBUTING FACTOR VEHICLE 3 CONTRIBUTING FACTOR VEHICLE 4  \
50000                           NaN                           NaN   
50001                           NaN                           NaN   

      CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID VEHICLE TYPE CODE 1  \
50000                           NaN       4244625               Sedan   
50001                           NaN       4242981               Sedan   

                       VEHICLE TYPE CODE 2 VEHICLE TYPE CODE 3  \
50000  Station Wagon/Sport Utility Vehicle                 NaN   
50001                            Box Truck                 NaN   

      VEHICLE TYPE CODE 4 VEHICLE TYPE CODE 5  
50000                 NaN                 NaN  
50001                 NaN                 NaN

In [11]:
df[df.duplicated()].tail(3)

DATETIME    BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
1612175 2012-07-09 14:35:00        NaN      NaN  40.791492 -73.935760   
1612176 2012-07-18 09:20:00  MANHATTAN    10024  40.779480 -73.973572   
1612177 2012-07-12 03:10:00     QUEENS    11361  40.762501 -73.782524   

                                                  LOCATION  \
1612175  {'type': 'Point', 'coordinates': [-73.935763, ...   
1612176  {'type': 'Point', 'coordinates': [-73.9735729,...   
1612177  {'type': 'Point', 'coordinates': [-73.7825216,...   

                           ON STREET NAME                 CROSS STREET NAME  \
1612175                               NaN                               NaN   
1612176  CENTRAL PARK WEST                 WEST 77 STREET                     
1612177  39 AVENUE                         203 STREET                         

        OFF STREET NAME  NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
1612175             NaN                        0.0                       0.0   
1612176             NaN                        0.0                       0.0   
1612177             NaN                        0.0                       0.0   

         NUMBER OF PEDESTRIANS INJURED  NUMBER OF PEDESTRIANS KILLED  \
1612175                              0                             0   
1612176                              0                             0   
1612177                              0                             0   

         NUMBER OF CYCLIST INJURED  NUMBER OF CYCLIST KILLED  \
1612175                          0                         0   
1612176                          0                         0   
1612177                          0                         0   

         NUMBER OF MOTORIST INJURED  NUMBER OF MOTORIST KILLED  \
1612175                           0                          0   
1612176                           0                          0   
1612177                           0                          0   

          CONTRIBUTING FACTOR VEHICLE 1 CONTRIBUTING FACTOR VEHICLE 2  \
1612175                 Fatigued/Drowsy                   Unspecified   
1612176  Driver Inattention/Distraction                   Unspecified   
1612177                     Unspecified                           NaN   

        CONTRIBUTING FACTOR VEHICLE 3 CONTRIBUTING FACTOR VEHICLE 4  \
1612175                           NaN                           NaN   
1612176                   Unspecified                           NaN   
1612177                           NaN                           NaN   

        CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID  \
1612175                           NaN       2853431   
1612176                           NaN         52405   
1612177                           NaN        259445   

                   VEHICLE TYPE CODE 1 VEHICLE TYPE CODE 2  \
1612175              PASSENGER VEHICLE   PASSENGER VEHICLE   
1612176  SPORT UTILITY / STATION WAGON                 VAN   
1612177              PASSENGER VEHICLE                 NaN   

        VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4 VEHICLE TYPE CODE 5  
1612175                 NaN                 NaN                 NaN  
1612176                 BUS                 NaN                 NaN  
1612177                 NaN                 NaN                 NaN

In [12]:
df[df['COLLISION_ID'].isin([4244625, 4242981, 2853431])]

DATETIME   BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
49726   2019-11-20 15:40:00    QUEENS    11101  40.744129 -73.927055   
49727   2019-11-12 07:05:00  BROOKLYN    11208  40.675938 -73.868233   
50000   2019-11-20 15:40:00    QUEENS    11101  40.744129 -73.927055   
50001   2019-11-12 07:05:00  BROOKLYN    11208  40.675938 -73.868233   
832914  2012-07-09 14:35:00       NaN      NaN  40.791492 -73.935760   
1612175 2012-07-09 14:35:00       NaN      NaN  40.791492 -73.935760   

                                                  LOCATION  \
49726    {'type': 'Point', 'coordinates': [-73.927055, ...   
49727    {'type': 'Point', 'coordinates': [-73.86823, 4...   
50000    {'type': 'Point', 'coordinates': [-73.927055, ...   
50001    {'type': 'Point', 'coordinates': [-73.86823, 4...   
832914   {'type': 'Point', 'coordinates': [-73.935763, ...   
1612175  {'type': 'Point', 'coordinates': [-73.935763, ...   

                           ON STREET NAME CROSS STREET NAME OFF STREET NAME  \
49726    QUEENS BOULEVARD                         38 STREET             NaN   
49727    PITKIN AVENUE                        AUTUMN AVENUE             NaN   
50000    QUEENS BOULEVARD                         38 STREET             NaN   
50001    PITKIN AVENUE                        AUTUMN AVENUE             NaN   
832914                                NaN               NaN             NaN   
1612175                               NaN               NaN             NaN   

         NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
49726                          0.0                       0.0   
49727                          0.0                       0.0   
50000                          0.0                       0.0   
50001                          0.0                       0.0   
832914                         0.0                       0.0   
1612175                        0.0                       0.0   

         NUMBER OF PEDESTRIANS INJURED  NUMBER OF PEDESTRIANS KILLED  \
49726                                0                             0   
49727                                0                             0   
50000                                0                             0   
50001                                0                             0   
832914                               0                             0   
1612175                              0                             0   

         NUMBER OF CYCLIST INJURED  NUMBER OF CYCLIST KILLED  \
49726                            0                         0   
49727                            0                         0   
50000                            0                         0   
50001                            0                         0   
832914                           0                         0   
1612175                          0                         0   

         NUMBER OF MOTORIST INJURED  NUMBER OF MOTORIST KILLED  \
49726                             0                          0   
49727                             0                          0   
50000                             0                          0   
50001                             0                          0   
832914                            0                          0   
1612175                           0                          0   

          CONTRIBUTING FACTOR VEHICLE 1   CONTRIBUTING FACTOR VEHICLE 2  \
49726    Driver Inattention/Distraction  Driver Inattention/Distraction   
49727                       Unspecified                     Unspecified   
50000    Driver Inattention/Distraction  Driver Inattention/Distraction   
50001                       Unspecified                     Unspecified   
832914                  Fatigued/Drowsy                     Unspecified   
1612175                 Fatigued/Drowsy                     Unspecified   

        CONTRIBUTING FACTOR VEHICLE 3 CONTRIBUTING FACTOR VEHICLE 4  \
49726                             NaN                         

In [13]:
df = df.drop_duplicates()

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1217957 entries, 0 to 1612019
Data columns (total 28 columns):
 #   Column                         Non-Null Count    Dtype         
---  ------                         --------------    -----         
 0   DATETIME                       1217957 non-null  datetime64[ns]
 1   BOROUGH                        847386 non-null   category      
 2   ZIP CODE                       847250 non-null   object        
 3   LATITUDE                       1071446 non-null  float32       
 4   LONGITUDE                      1071446 non-null  float32       
 5   LOCATION                       1071446 non-null  category      
 6   ON STREET NAME                 977625 non-null   object        
 7   CROSS STREET NAME              804464 non-null   object        
 8   OFF STREET NAME                171749 non-null   object        
 9   NUMBER OF PERSONS INJURED      1217944 non-null  float32       
 10  NUMBER OF PERSONS KILLED       1217932 non-null  float32   

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.3 Handling missing BOROUGH

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.3.1 Checking BOROUGH coulumn

In [15]:
df_len = len(df)
print(f'Dataset entries: {df_len}')

missing_borough_sum = df['BOROUGH'].isnull().sum()
print(f'Missing BOROUGH: {missing_borough_sum}')

missing_borough_percentage = (missing_borough_sum / df_len) * 100
print(f"Missing BOROUGH make up {missing_borough_percentage:.2f}% of the main dataset.")

Dataset entries: 1217957
Missing BOROUGH: 370571
Missing BOROUGH make up 30.43% of the main dataset.


New York City (NYC) is composed of five boroughs: Manhattan, Brooklyn, Queens, The Bronx, and Staten Island.

In [16]:
df['BOROUGH'].value_counts()

BOROUGH
BROOKLYN         263285
QUEENS           225555
MANHATTAN        203102
BRONX            118701
STATEN ISLAND     36743
Name: count, dtype: int64

In [17]:
valid_boroughs = ['BROOKLYN', 'QUEENS', 'MANHATTAN', 'BRONX', 'STATEN ISLAND']

invalid_boroughs = df[~df['BOROUGH'].isin(valid_boroughs)]
invalid_boroughs['BOROUGH'].unique()

[NaN]
Categories (5, object): ['BRONX', 'BROOKLYN', 'MANHATTAN', 'QUEENS', 'STATEN ISLAND']

 Provided values are clean and correct.

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.3.2 Checking columns providing localization

In [18]:
mask = df['ZIP CODE'].notnull() & df['BOROUGH'].isnull()

values_provided_for_zipcode = df[mask].shape[0]
print(f'Number of rows that could potentially be filled using ZIP CODE: {values_provided_for_zipcode}')

Number of rows that could potentially be filled using ZIP CODE: 0


When BOROUGH is NaN then ZIP CODE is NaN too.

Decided to drop ZIP CODE column as is no longer neccesary.

In [19]:
df.drop(columns=['ZIP CODE'], inplace=True)

In [20]:
mask = df['BOROUGH'].isnull() & df['LATITUDE'].notnull() & df['LONGITUDE'].notnull() & df['LOCATION'].notnull()

values_provided_for_lat_long_loc = df[mask].shape[0]
values_provided_for_lat_long_loc_percentage = (values_provided_for_lat_long_loc / missing_borough_sum) * 100
print(f'Rows with missing BOROUGH that could potentially be filled using LATITUDE, LONGITUDE, or LOCATION: {values_provided_for_lat_long_loc}, which make up {values_provided_for_lat_long_loc_percentage:.2f}% of missing BOROUGH.')


Rows with missing BOROUGH that could potentially be filled using LATITUDE, LONGITUDE, or LOCATION: 246199, which make up 66.44% of missing BOROUGH.


In [21]:
mask = df['BOROUGH'].isnull() & df['LATITUDE'].isnull() & df['LONGITUDE'].isnull() & df['LOCATION'].notnull()

values_provided_only_for_loc = df[mask].shape[0]
print(f'Number of rows that have LOCATION only: {values_provided_only_for_loc}')

Number of rows that have LOCATION only: 0


If the LOCATION is given, then LONGITUDE and LATITUDE are also provided.

Verifying whether the LONGITUDE and LATITUDE columns contain the same coordinate values as the LOCATION column:

In [22]:
pd.set_option('display.precision', 10)

In [23]:
longitude_values = df['LONGITUDE']
location_longitude_values = df['LOCATION'].str.extract(r'\[([^\]]+)\]')[0].str.split(',', expand=True)[0].str.strip().astype(float)

print(longitude_values.head(3))
print(location_longitude_values.head(3))
print(longitude_values == location_longitude_values)

0   -73.7901840210
1   -73.9510955811
2   -73.9938354492
Name: LONGITUDE, dtype: float32
0   -73.790184
1   -73.951096
2   -73.993835
Name: 0, dtype: float64
0          False
1          False
2          False
3          False
4          False
           ...  
1611230    False
1611231    False
1611232    False
1611233    False
1612019    False
Length: 1217957, dtype: bool


The comparison above confirms that the columns contain the same coordinate values, with the LONGITUDE and LATITUDE columns providing more decimal places (and therefore higher accuracy). 

The approximate accuracy of geographic coordinates depends on the number of decimal places:
- 6 decimal places provide an accuracy of ~0.11 meters
- The difference for two such values can be around ~15 cm

In this context, an accuracy of ~110 meters (corresponding to 3 decimal places) would be sufficient for general location analysis.

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.3.3 Attempt to infer the missing borough values using LATITUDE and LONGITUDE.

In [24]:
import geopandas as gpd
from shapely.geometry import Point

In [25]:
NYC_boroughs = gpd.read_file("NYC_Borough_Boundary.geojson")
NYC_boroughs = NYC_boroughs.to_crs(epsg=4326)

print(f'Column names in GeoDataFrame: {NYC_boroughs.columns}')

Column names in GeoDataFrame: Index(['OBJECTID', 'BoroCode', 'BoroName', 'geometry'], dtype='object')


In [26]:
# Filter rows with missing BOROUGH and valid LAT/LON
mask = df['BOROUGH'].isna() & df['LATITUDE'].notna() & df['LONGITUDE'].notna()
df_missing_borough = df[mask].copy()
df_missing_borough.info()

<class 'pandas.core.frame.DataFrame'>
Index: 246199 entries, 7 to 1611232
Data columns (total 27 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   DATETIME                       246199 non-null  datetime64[ns]
 1   BOROUGH                        0 non-null       category      
 2   LATITUDE                       246199 non-null  float32       
 3   LONGITUDE                      246199 non-null  float32       
 4   LOCATION                       246199 non-null  category      
 5   ON STREET NAME                 190917 non-null  object        
 6   CROSS STREET NAME              33627 non-null   object        
 7   OFF STREET NAME                9577 non-null    object        
 8   NUMBER OF PERSONS INJURED      246196 non-null  float32       
 9   NUMBER OF PERSONS KILLED       246195 non-null  float32       
 10  NUMBER OF PEDESTRIANS INJURED  246199 non-null  int8          
 11  NUMB

In [27]:
# Create geometry column
geometry = [Point(xy) for xy in zip(df_missing_borough['LONGITUDE'], df_missing_borough['LATITUDE'])]
gdf_missing_borough = gpd.GeoDataFrame(df_missing_borough, geometry=geometry, crs='EPSG:4326')

# Perform spatial join to assign boroughs
df_borough_updated = gpd.sjoin(gdf_missing_borough, NYC_boroughs[['BoroName', 'geometry']], how='left', predicate='within')

# Convert to uppercase to match existing BOROUGH categories
df_borough_updated['BoroName'] = df_borough_updated['BoroName'].str.upper()

# Fill in the BOROUGH values back to the main df
df.loc[df_borough_updated.index, 'BOROUGH'] = df_borough_updated['BoroName']

missing_borough_values_filled = df_borough_updated['BoroName'].notna().sum()

missing_borough_values_filled_percentage = (missing_borough_values_filled / values_provided_for_lat_long_loc) * 100

print(f"Number of filled missing boroughs using coordinates: {missing_borough_values_filled}, which make up {missing_borough_values_filled_percentage:.2f}% of missing BOROUGH that could potentially be filled using LATITUDE, LONGITUDE, or LOCATION.")

Number of filled missing boroughs using coordinates: 240152, which make up 97.54% of missing BOROUGH that could potentially be filled using LATITUDE, LONGITUDE, or LOCATION.


In [28]:
borough_values_not_filled = df_borough_updated['BoroName'].isna().sum()
missing_borough_percentage = (df_borough_updated['BoroName'].isna().sum() / df_len) * 100
missing_borough_values_percentage = (borough_values_not_filled / values_provided_for_lat_long_loc) * 100

print(f"Number of missing BOROUGH values: {borough_values_not_filled}, which make up {missing_borough_values_percentage:.2f}% of missing BOROUGH that could potentially be filled using LATITUDE, LONGITUDE, or LOCATION and {missing_borough_percentage:.2f}% of the main dataset.")

Number of missing BOROUGH values: 6047, which make up 2.46% of missing BOROUGH that could potentially be filled using LATITUDE, LONGITUDE, or LOCATION and 0.50% of the main dataset.


Checking if the result is accurate:

In [29]:
expected_unfilled = values_provided_for_lat_long_loc - missing_borough_values_filled
actual_unfilled = borough_values_not_filled

result = (expected_unfilled == actual_unfilled)

print(f'The missing BOROUGH value is consistent: {result}')

The missing BOROUGH value is consistent: True


In [30]:
print("---\nSummary:\n---")
print(f'Missing BOROUGH: {missing_borough_sum}')
print(f'Number of rows that could potentially be filled using LATITUDE, LONGITUDE or LOCATION: {values_provided_for_lat_long_loc}')
print(f"Number of filled missing boroughs using coordinates: {missing_borough_values_filled}")
print(f"Number of missing BOROUGH values: {borough_values_not_filled}, make up {missing_borough_percentage:.2f}% of the main dataset.")
print('---')

missing_borough_in_main = df['BOROUGH'].isna().sum()

print(f'Still missing values in main df: {missing_borough_in_main}\n---')

---
Summary:
---
Missing BOROUGH: 370571
Number of rows that could potentially be filled using LATITUDE, LONGITUDE or LOCATION: 246199
Number of filled missing boroughs using coordinates: 240152
Number of missing BOROUGH values: 6047, make up 0.50% of the main dataset.
---
Still missing values in main df: 130419
---


#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.3.4 Attempt to infer missing BOROUGH values using LOCATION (likely outside NYC)

In [31]:
missing_borough_main = df['BOROUGH'].isna().sum()
count_missing_borough = df[df['BOROUGH'].isna() & df['LOCATION'].notna()].shape[0]

print(f'Number of missing BOROUGH values in main df: {missing_borough_main}')
print(f'Number of rows where BOROUGH is missing and LOCATION is provided: {count_missing_borough}')

Number of missing BOROUGH values in main df: 130419
Number of rows where BOROUGH is missing and LOCATION is provided: 6047


In [32]:
outside_nyc = df[df['BOROUGH'].isna() & df['LOCATION'].notna()].copy()

# Extract coordinates
outside_nyc['COORDINATES'] = outside_nyc['LOCATION'].str.extract(r'\[([^\]]+)\]')[0]

# Split and convert to float
split_coords = outside_nyc['COORDINATES'].str.split(',', expand=True)
outside_nyc['LON'] = split_coords[0].str.strip().astype(float)
outside_nyc['LAT'] = split_coords[1].str.strip().astype(float)

In [33]:
print(outside_nyc['LON'].head(5))
print(outside_nyc['LAT'].head(5))

55     -73.96012
959    -73.92913
2762   -73.91276
2847   -73.92938
3371   -73.92813
Name: LON, dtype: float64
55      40.756367
959     40.727870
2762    40.585710
2847    40.727660
3371    40.845467
Name: LAT, dtype: float64


In [34]:
# Create geometry column
geometry = [Point(xy) for xy in zip(outside_nyc['LON'], outside_nyc['LAT'])]
gdf_missing = gpd.GeoDataFrame(outside_nyc, geometry=geometry, crs='EPSG:4326')

# Perform spatial join to assign boroughs
df_borough_updated = gpd.sjoin(gdf_missing, NYC_boroughs[['BoroName', 'geometry']], how='left', predicate='within')

# Convert to uppercase to match existing BOROUGH categories
df_borough_updated['BoroName'] = df_borough_updated['BoroName'].str.upper()

# Fill in the BOROUGH values back to the main DataFrame
df.loc[df_borough_updated.index, 'BOROUGH'] = df_borough_updated['BoroName']

missing_borough_values_filled = df_borough_updated['BoroName'].notna().sum()

# Done — preview how many were filled
print(f"Filled {df_borough_updated['BoroName'].notna().sum()} missing boroughs using location coordinates.")

Filled 0 missing boroughs using location coordinates.


Using coordinates from LOCATION didn't fill up any additional BOROUGH values (6047 missing data, make up 0.50% of the main dataset). 

LOCATION provides the same values as LONGITUDE and LATITUDE, so LONGITUDE and LATITUDE were dropped.

In [35]:
df.drop(columns=['LATITUDE', 'LONGITUDE'], inplace=True)

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.3.5 Attempt to infer the missing BOROUGH values using streets data

Checking the next column that is most likely to help fill the missing BOROUGH values

In [36]:
df[['BOROUGH', 'ON STREET NAME', 'CROSS STREET NAME', 'OFF STREET NAME']].isnull().sum()

BOROUGH               130419
ON STREET NAME        240332
CROSS STREET NAME     413493
OFF STREET NAME      1046208
dtype: int64

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.3.5.1 Attempt to fill missing BOROUGH values using ON STREET NAME

In [37]:
missing_borough_main = df['BOROUGH'].isna().sum()
count_missing_borough = df[df['BOROUGH'].isna() & df['ON STREET NAME'].notna()].shape[0]
missing_borough_percentage = (count_missing_borough / missing_borough_main) * 100
missing_borough_percentage_main = (count_missing_borough / df_len) * 100

print(f'---\nNumber of missing BOROUGH values in main df: {missing_borough_main}')
print(f'Number of rows where BOROUGH is missing and ON STREET NAME is provided: {count_missing_borough}\n---')

print(f"Missing BOROUGH with provided ON STREET NAME make up {missing_borough_percentage:.2f}% of the missing value for BOROUGH.")
print(f"Missing BOROUGH with provided ON STREET NAME make up {missing_borough_percentage_main:.2f}% of the main dataset.\n---")

---
Number of missing BOROUGH values in main df: 130419
Number of rows where BOROUGH is missing and ON STREET NAME is provided: 94114
---
Missing BOROUGH with provided ON STREET NAME make up 72.16% of the missing value for BOROUGH.
Missing BOROUGH with provided ON STREET NAME make up 7.73% of the main dataset.
---


In [38]:
df_missing_borough_onstreetname = df[df['BOROUGH'].isna() & df['ON STREET NAME'].notna()]

In [39]:
df_missing_borough_onstreetname.info()

<class 'pandas.core.frame.DataFrame'>
Index: 94114 entries, 26 to 1609980
Data columns (total 25 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   DATETIME                       94114 non-null  datetime64[ns]
 1   BOROUGH                        0 non-null      category      
 2   LOCATION                       5467 non-null   category      
 3   ON STREET NAME                 94114 non-null  object        
 4   CROSS STREET NAME              72957 non-null  object        
 5   OFF STREET NAME                30 non-null     object        
 6   NUMBER OF PERSONS INJURED      94113 non-null  float32       
 7   NUMBER OF PERSONS KILLED       94112 non-null  float32       
 8   NUMBER OF PEDESTRIANS INJURED  94114 non-null  int8          
 9   NUMBER OF PEDESTRIANS KILLED   94114 non-null  int8          
 10  NUMBER OF CYCLIST INJURED      94114 non-null  int8          
 11  NUMBER OF CYCLIST

In [40]:
df_missing_borough_onstreetname['ON STREET NAME'].unique()

array(['WESTCHESTER AVENUE              ',
       'HUTCHINSON RIVER PARKWAY        ',
       'VANWYCK EXPRESSWAY              ', ...,
       'ELK STREET                      ',
       'BQE WEST SOUTH SR               ',
       'MORROW STREET                   '], shape=(5087,), dtype=object)

In [41]:
df_missing_borough_onstreetname = df_missing_borough_onstreetname.copy()

# Cleaning the street names
df_missing_borough_onstreetname.loc[:, 'ON STREET NAME CLEAN'] = df_missing_borough_onstreetname['ON STREET NAME'].str.strip().str.upper()

# Counting the number of occurrences
street_counts = df_missing_borough_onstreetname['ON STREET NAME CLEAN'].value_counts()

print(street_counts.head(10))

ON STREET NAME CLEAN
HORACE HARDING EXPRESSWAY    1751
FLATBUSH AVENUE              1263
VERRAZANO BRIDGE UPPER       1251
BRUCKNER BOULEVARD           1152
BRONX WHITESTONE BRIDGE      1117
QUEENS BOULEVARD              995
NORTHERN BOULEVARD            968
BELT PARKWAY                  966
RICHMOND AVENUE               929
ROCKAWAY BOULEVARD            905
Name: count, dtype: int64


In [42]:
# Filter street names with counts >= 100
high_freq_streets = street_counts[street_counts >= 100]

# Get the last index (least frequent among those >= 100)
last_index_position = street_counts.index.get_loc(high_freq_streets.index[-1])
print(last_index_position)

195


Creating top_street_names.txt with street names that occur more than 100 times in the dataset:

In [43]:
top_street_names = street_counts.iloc[:196]

# Save only the street names (index) to a .txt file
with open("top_street_names_onstreetname.txt", "w", encoding="utf-8") as f:
    for street_name in top_street_names.index:
        f.write(f"{street_name}\n")

Based on top_street_names.txt, the script street_to_borough.py labels each street by its borough. Streets that occur in more than one borough were dropped.

Filling the BOROUGH column using labeled data from the borough dictionary in street_to_borough.py:

In [44]:
from street_to_borough import borough

df_missing_borough_onstreetname['BOROUGH_FROM_STREET'] = df_missing_borough_onstreetname['ON STREET NAME CLEAN'].map(borough)
df_missing_borough_onstreetname.loc[df_missing_borough_onstreetname['BOROUGH'].isna(), 'BOROUGH'] = \
    df_missing_borough_onstreetname.loc[df_missing_borough_onstreetname['BOROUGH'].isna(), 'BOROUGH_FROM_STREET']
df_missing_borough_onstreetname.drop(columns=['BOROUGH_FROM_STREET', 'ON STREET NAME CLEAN'], inplace=True)

# Count missing after
missing_after = df_missing_borough_onstreetname['BOROUGH'].isna().sum()

# Calculate how many were filled
filled_count = count_missing_borough - missing_after
print(f"Number of borough values filled using street names: {filled_count}")

Number of borough values filled using street names: 57414


In [45]:
df.loc[df_missing_borough_onstreetname.index, 'BOROUGH'] = df_missing_borough_onstreetname['BOROUGH']

In [46]:
missing_borough_main = df['BOROUGH'].isna().sum()
missing_borough_percentage = (filled_count / count_missing_borough) * 100
missing_borough_percentage_main = (missing_borough_main / df_len) * 100

print(f'---\nNumber of missing BOROUGH values in main df: {missing_borough_main}')
print(f"Borough filled using street name: {filled_count}\n---")

print(f"Filled BOROUGH using ON STREET NAME make up {missing_borough_percentage:.2f}% of the missing values for BOROUGH.")
print(f"Missing BOROUGH make up {missing_borough_percentage_main:.2f}% of the main dataset.\n---")

---
Number of missing BOROUGH values in main df: 73005
Borough filled using street name: 57414
---
Filled BOROUGH using ON STREET NAME make up 61.00% of the missing values for BOROUGH.
Missing BOROUGH make up 5.99% of the main dataset.
---


#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.3.5.2 Verifying if other street-related data can help complete the missing BOROUGH information.

In [47]:
df['CROSS STREET NAME'].unique()

array([nan, 'GRAND STREET', '2 AVENUE', ...,
       'BOULDER STREET                  ',
       'JUPITER LANE                    ',
       'SUTHERLAND STREET               '], shape=(16870,), dtype=object)

In [48]:
df_missing_borough_crossstreetname = df[df['BOROUGH'].isna() & df['CROSS STREET NAME'].notna()]

In [49]:
count_missing_borough = df_missing_borough_crossstreetname['BOROUGH'].isna().sum()
missing_borough_percentage = (count_missing_borough / missing_borough_main) * 100
missing_borough_percentage_main = (count_missing_borough / df_len) * 100

print(f'---\nNumber of missing BOROUGH values in main df: {missing_borough_main}')
print(f'Number of rows where BOROUGH is missing and CROSS STREET NAME is provided: {count_missing_borough}\n---')

print(f"Missing BOROUGH with provided CROSS STREET NAME make up {missing_borough_percentage:.2f}% of the missing value for BOROUGH.")
print(f"Missing BOROUGH with provided CROSS STREET NAME make up {missing_borough_percentage_main:.2f}% of the main dataset.\n---")

---
Number of missing BOROUGH values in main df: 73005
Number of rows where BOROUGH is missing and CROSS STREET NAME is provided: 33345
---
Missing BOROUGH with provided CROSS STREET NAME make up 45.67% of the missing value for BOROUGH.
Missing BOROUGH with provided CROSS STREET NAME make up 2.74% of the main dataset.
---


In [50]:
df_missing_borough_crossstreetname = df_missing_borough_crossstreetname.copy()

# Cleaning the street names
df_missing_borough_crossstreetname.loc[:, 'CROSS STREET NAME CLEAN'] = df_missing_borough_crossstreetname['CROSS STREET NAME'].str.strip().str.upper()

# Counting the number of occurrences
street_counts = df_missing_borough_crossstreetname['CROSS STREET NAME CLEAN'].value_counts()

print(street_counts.head(10))

CROSS STREET NAME CLEAN
SHORE PARKWAY                531
BROADWAY                     449
HORACE HARDING EXPRESSWAY    433
BRUCKNER BOULEVARD           414
CROSS BRONX EXPRESSWAY       381
GRAND CENTRAL PARKWAY        379
PARK AVENUE                  362
JAMAICA AVENUE               309
WHITESTONE EXPRESSWAY        301
3 AVENUE                     260
Name: count, dtype: int64


In [51]:
# Filter street names with counts >= 100
high_freq_streets = street_counts[street_counts >= 100]

# Get the last index (least frequent among those >= 100)
last_index_position = street_counts.index.get_loc(high_freq_streets.index[-1])
print(last_index_position)

40


In [52]:
top_street_names = street_counts.iloc[:41]

# Save only the street names (index) to a .txt file
with open("top_street_names_crossstreetname.txt", "w", encoding="utf-8") as f:
    for street_name in top_street_names.index:
        f.write(f"{street_name}\n")

In [53]:
from street_to_borough_crossnames import borough_from_cross_streets

df_missing_borough_crossstreetname['BOROUGH_FROM_CROSS_STREET'] = df_missing_borough_crossstreetname['CROSS STREET NAME CLEAN'].map(borough_from_cross_streets)
df_missing_borough_crossstreetname.loc[df_missing_borough_crossstreetname['BOROUGH'].isna(), 'BOROUGH'] = \
    df_missing_borough_crossstreetname.loc[df_missing_borough_crossstreetname['BOROUGH'].isna(), 'BOROUGH_FROM_CROSS_STREET']
df_missing_borough_crossstreetname.drop(columns=['BOROUGH_FROM_CROSS_STREET', 'CROSS STREET NAME CLEAN'], inplace=True)

# Count missing after
missing_after = df_missing_borough_crossstreetname['BOROUGH'].isna().sum()

# Calculate how many were filled
filled_count = count_missing_borough - missing_after
print(f"Borough filled using cross street name: {filled_count}")

Borough filled using cross street name: 6128


In [54]:
df.loc[df_missing_borough_crossstreetname.index, 'BOROUGH'] = df_missing_borough_crossstreetname['BOROUGH']

In [55]:
missing_borough_main = df['BOROUGH'].isna().sum()
missing_borough_percentage = (filled_count / count_missing_borough) * 100
missing_borough_percentage_main = (missing_borough_main / df_len) * 100

print(f'---\nNumber of missing BOROUGH values in main df: {missing_borough_main}')
print(f"Borough filled using CROSS STREET NAME: {filled_count}\n---")

print(f"Filled BOROUGH using CROSS STREET NAME make up {missing_borough_percentage:.2f}% of the missing values for BOROUGH.")
print(f"Missing BOROUGH make up {missing_borough_percentage_main:.2f}% of the main dataset.\n---")

---
Number of missing BOROUGH values in main df: 66877
Borough filled using CROSS STREET NAME: 6128
---
Filled BOROUGH using CROSS STREET NAME make up 18.38% of the missing values for BOROUGH.
Missing BOROUGH make up 5.49% of the main dataset.
---


In [56]:
df_missing_borough_offstreetname = df[df['BOROUGH'].isna() & df['OFF STREET NAME'].notna()]
df_missing_borough_offstreetname['BOROUGH'].isna().sum()

np.int64(13177)

In [57]:
df_missing_borough_offstreetname['OFF STREET NAME'].value_counts()

OFF STREET NAME
PARKING LOT 110-00 ROCKAWAY BOULEVARD       86
PARKING LOT-772 EDGEWATER RD                61
PARKING LOT OF 110-00 ROCKAWAY BOULEVARD    53
8973 BAY PKWY (PARKING LOT)                 41
MUNI LOT 1                                  37
                                            ..
PKG LOT OF 46-31 KISSENA BLVD                1
1127 EAST 57 STREET(BRICK WALL)              1
GAS STATION 119-01 ATLANTIC AVE              1
SIDEWALK 105-36 90 STREET                    1
GAS STATION 166-06 HILLSIDE AVE              1
Name: count, Length: 9762, dtype: int64

Missing BOROUGH values make up 5,49% of the main dataset. 

The provided data in the OFF STREET NAME column is not sufficient to efficiently fill the missing values in the BOROUGH column. Missing values in the BOROUGH column account for 5.49% of the dataset, of which 0.5% likely correspond to locations outside of NYC. Therefore, the decision was made to drop the remaining rows with missing borough values.

In [58]:
df = df.dropna(subset=['BOROUGH'])

Creating a column STREET that contains the street information extracted from other street-related columns, which will be dropped afterward.

In [59]:
# Creating one column containing the thata from street-related columns
df['STREET'] = df['ON STREET NAME'].fillna(df['CROSS STREET NAME']).fillna(df['OFF STREET NAME'])

# Get the column list
cols = df.columns.tolist()

# Remove 'STREET' from the list and reinsert it at index 3
cols.remove('STREET')
cols.insert(3, 'STREET')

# Reorder the DataFrame
df = df[cols]

In [60]:
# dropping columns: ON STREET NAME, CROSS STREET NAME, OFF STREAT NAME
df.drop(columns=['ON STREET NAME', 'CROSS STREET NAME', 'OFF STREET NAME'], inplace=True)

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1151080 entries, 0 to 1611233
Data columns (total 23 columns):
 #   Column                         Non-Null Count    Dtype         
---  ------                         --------------    -----         
 0   DATETIME                       1151080 non-null  datetime64[ns]
 1   BOROUGH                        1151080 non-null  category      
 2   LOCATION                       1070515 non-null  category      
 3   STREET                         1105463 non-null  object        
 4   NUMBER OF PERSONS INJURED      1151068 non-null  float32       
 5   NUMBER OF PERSONS KILLED       1151057 non-null  float32       
 6   NUMBER OF PEDESTRIANS INJURED  1151080 non-null  int8          
 7   NUMBER OF PEDESTRIANS KILLED   1151080 non-null  int8          
 8   NUMBER OF CYCLIST INJURED      1151080 non-null  int8          
 9   NUMBER OF CYCLIST KILLED       1151080 non-null  int8          
 10  NUMBER OF MOTORIST INJURED     1151080 non-null  int8      

In [62]:
df.isna().sum()

DATETIME                               0
BOROUGH                                0
LOCATION                           80565
STREET                             45617
NUMBER OF PERSONS INJURED             12
NUMBER OF PERSONS KILLED              23
NUMBER OF PEDESTRIANS INJURED          0
NUMBER OF PEDESTRIANS KILLED           0
NUMBER OF CYCLIST INJURED              0
NUMBER OF CYCLIST KILLED               0
NUMBER OF MOTORIST INJURED             0
NUMBER OF MOTORIST KILLED              0
CONTRIBUTING FACTOR VEHICLE 1       3131
CONTRIBUTING FACTOR VEHICLE 2     155745
CONTRIBUTING FACTOR VEHICLE 3    1077359
CONTRIBUTING FACTOR VEHICLE 4    1135704
CONTRIBUTING FACTOR VEHICLE 5    1147105
COLLISION_ID                           0
VEHICLE TYPE CODE 1                 4186
VEHICLE TYPE CODE 2               195562
VEHICLE TYPE CODE 3              1059188
VEHICLE TYPE CODE 4              1117601
VEHICLE TYPE CODE 5              1143777
dtype: int64

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.4 Handling missing NUMBER OF PERSONS INJURED/KILLED

In [63]:
# Filling NaN values with 0
df['NUMBER OF PERSONS INJURED'] = df['NUMBER OF PERSONS INJURED'].fillna(0)
df['NUMBER OF PERSONS KILLED'] = df['NUMBER OF PERSONS KILLED'].fillna(0)

# Convert data types to int
df['NUMBER OF PERSONS INJURED'] = df['NUMBER OF PERSONS INJURED'].astype('int8')
df['NUMBER OF PERSONS KILLED'] = df['NUMBER OF PERSONS KILLED'].astype('int8')

In [64]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1151080 entries, 0 to 1611233
Data columns (total 23 columns):
 #   Column                         Non-Null Count    Dtype         
---  ------                         --------------    -----         
 0   DATETIME                       1151080 non-null  datetime64[ns]
 1   BOROUGH                        1151080 non-null  category      
 2   LOCATION                       1070515 non-null  category      
 3   STREET                         1105463 non-null  object        
 4   NUMBER OF PERSONS INJURED      1151080 non-null  int8          
 5   NUMBER OF PERSONS KILLED       1151080 non-null  int8          
 6   NUMBER OF PEDESTRIANS INJURED  1151080 non-null  int8          
 7   NUMBER OF PEDESTRIANS KILLED   1151080 non-null  int8          
 8   NUMBER OF CYCLIST INJURED      1151080 non-null  int8          
 9   NUMBER OF CYCLIST KILLED       1151080 non-null  int8          
 10  NUMBER OF MOTORIST INJURED     1151080 non-null  int8      

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.4.1 Verifying whether the NUMBER OF PERSONS INJURED and NUMBER OF PERSONS KILLED columns represent the total counts of pedestrians, cyclists, and motorists.

In [65]:
# Check for discrepancies in NUMBER OF PERSONS INJURED
injured_missing = (df['NUMBER OF PERSONS INJURED'] != (
                       df['NUMBER OF PEDESTRIANS INJURED'] + 
                       df['NUMBER OF CYCLIST INJURED'] + 
                       df['NUMBER OF MOTORIST INJURED']))

# Check for discrepancies in NUMBER OF PERSONS KILLED
killed_missing = (df['NUMBER OF PERSONS KILLED'] != (
                       df['NUMBER OF PEDESTRIANS KILLED'] + 
                       df['NUMBER OF CYCLIST KILLED'] + 
                       df['NUMBER OF MOTORIST KILLED']))

# Combine both conditions and filter the rows with discrepancies
injured_and_killed_missing = df[injured_missing & killed_missing].shape[0]
injured_killed_missing = df[injured_missing | killed_missing].shape[0]

print(f'Number of rows where the total number of injured OR killed differs from the sum of pedestrians, cyclists, and motorists: {injured_killed_missing}.')
print(f'Number of rows where the total number of injured AND killed differs from the sum of pedestrians, cyclists, and motorists: {injured_and_killed_missing}.')

Number of rows where the total number of injured OR killed differs from the sum of pedestrians, cyclists, and motorists: 291.
Number of rows where the total number of injured AND killed differs from the sum of pedestrians, cyclists, and motorists: 1.


In [66]:
df_len = df.shape[0]
injured_killed_missing_percentage = (injured_killed_missing / df_len) * 100

print(f'---\nNumber of dataset rows: {df_len}')
print(f'Percentage of differed values for injured/killed: {injured_killed_missing_percentage:.3f}%\n---')

---
Number of dataset rows: 1151080
Percentage of differed values for injured/killed: 0.025%
---


The number of discrepancies is small enough to be considered data errors. Therefore, treating the NUMBER OF PERSONS INJURED and NUMBER OF PERSONS KILLED as accurate totals seems reasonable.

In [67]:
# Calculate the component sums
sum_injured = df['NUMBER OF PEDESTRIANS INJURED'] + df['NUMBER OF CYCLIST INJURED'] + df['NUMBER OF MOTORIST INJURED']
sum_killed = df['NUMBER OF PEDESTRIANS KILLED'] + df['NUMBER OF CYCLIST KILLED'] + df['NUMBER OF MOTORIST KILLED']

# Masks for updates
mask_injured_update = (df['NUMBER OF PERSONS INJURED'] == 0) & (sum_injured > 0)
mask_killed_update = (df['NUMBER OF PERSONS KILLED'] == 0) & (sum_killed > 0)

# Perform updates
df.loc[mask_injured_update, 'NUMBER OF PERSONS INJURED'] = sum_injured
df.loc[mask_killed_update, 'NUMBER OF PERSONS KILLED'] = sum_killed

print(f"Updated NUMBER OF PERSONS INJURED from sum: {mask_injured_update.sum()}")
print(f"Updated NUMBER OF PERSONS KILLED from sum: {mask_killed_update.sum()}")

Updated NUMBER OF PERSONS INJURED from sum: 151
Updated NUMBER OF PERSONS KILLED from sum: 5


In [68]:
# Check for discrepancies in NUMBER OF PERSONS INJURED
injured_missing = (df['NUMBER OF PERSONS INJURED'] != (
                       df['NUMBER OF PEDESTRIANS INJURED'] + 
                       df['NUMBER OF CYCLIST INJURED'] + 
                       df['NUMBER OF MOTORIST INJURED']))

# Check for discrepancies in NUMBER OF PERSONS KILLED
killed_missing = (df['NUMBER OF PERSONS KILLED'] != (
                       df['NUMBER OF PEDESTRIANS KILLED'] + 
                       df['NUMBER OF CYCLIST KILLED'] + 
                       df['NUMBER OF MOTORIST KILLED']))

# Combine both conditions and filter the rows with discrepancies
injured_killed_missing = df[injured_missing | killed_missing].shape[0]

print(f'Number of rows where the total number of injured or killed differs from the sum of pedestrians, cyclists, and motorists: {injured_killed_missing}.')

Number of rows where the total number of injured or killed differs from the sum of pedestrians, cyclists, and motorists: 136.


In [69]:
injured_killed_missing_percentage = (injured_killed_missing / df_len) * 100

print(f'---\nNumber of dataset rows: {df_len}')
print(f'Percentage of differed values for injured/killed: {injured_killed_missing_percentage:.3f}%\n---')

---
Number of dataset rows: 1151080
Percentage of differed values for injured/killed: 0.012%
---


Decided to drop the remaining rows where the total number of injured or killed does not match the sum of pedestrians, cyclists, and motorists, as the data is likely inaccurate. These rows account for only 0.012% of the dataset, so their removal is not expected to significantly impact the overall analysis.

In [70]:
invalid_rows = injured_missing | killed_missing
df = df[~invalid_rows].copy()

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1150944 entries, 0 to 1611233
Data columns (total 23 columns):
 #   Column                         Non-Null Count    Dtype         
---  ------                         --------------    -----         
 0   DATETIME                       1150944 non-null  datetime64[ns]
 1   BOROUGH                        1150944 non-null  category      
 2   LOCATION                       1070392 non-null  category      
 3   STREET                         1105331 non-null  object        
 4   NUMBER OF PERSONS INJURED      1150944 non-null  int8          
 5   NUMBER OF PERSONS KILLED       1150944 non-null  int8          
 6   NUMBER OF PEDESTRIANS INJURED  1150944 non-null  int8          
 7   NUMBER OF PEDESTRIANS KILLED   1150944 non-null  int8          
 8   NUMBER OF CYCLIST INJURED      1150944 non-null  int8          
 9   NUMBER OF CYCLIST KILLED       1150944 non-null  int8          
 10  NUMBER OF MOTORIST INJURED     1150944 non-null  int8      

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.2.5 Handling CONTRIBUTING FACTOR VEHICLE 1-5 and VEHICLE TYPE CODE 1-5

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;1.2.5.1 Checking CONTRIBUTING FACTOR VEHICLE 1-5

In [72]:
df.isnull().sum()

DATETIME                               0
BOROUGH                                0
LOCATION                           80552
STREET                             45613
NUMBER OF PERSONS INJURED              0
NUMBER OF PERSONS KILLED               0
NUMBER OF PEDESTRIANS INJURED          0
NUMBER OF PEDESTRIANS KILLED           0
NUMBER OF CYCLIST INJURED              0
NUMBER OF CYCLIST KILLED               0
NUMBER OF MOTORIST INJURED             0
NUMBER OF MOTORIST KILLED              0
CONTRIBUTING FACTOR VEHICLE 1       3112
CONTRIBUTING FACTOR VEHICLE 2     155663
CONTRIBUTING FACTOR VEHICLE 3    1077230
CONTRIBUTING FACTOR VEHICLE 4    1135570
CONTRIBUTING FACTOR VEHICLE 5    1146970
COLLISION_ID                           0
VEHICLE TYPE CODE 1                 4170
VEHICLE TYPE CODE 2               195483
VEHICLE TYPE CODE 3              1059064
VEHICLE TYPE CODE 4              1117468
VEHICLE TYPE CODE 5              1143641
dtype: int64

In [73]:
mask = (
    df['CONTRIBUTING FACTOR VEHICLE 1'].isnull() & (
    df['CONTRIBUTING FACTOR VEHICLE 2'].notnull() |
    df['CONTRIBUTING FACTOR VEHICLE 3'].notnull() |
    df['CONTRIBUTING FACTOR VEHICLE 4'].notnull() |
    df['CONTRIBUTING FACTOR VEHICLE 5'].notnull())
)

first_factor_missing_other_in = df[mask].shape[0]
print(f'Number of rows where CONTRIBUTING FACTOR VEHICLE 1 is missing, but at least one of CONTRIBUTING FACTOR VEHICLE 2-5 is provided: {first_factor_missing_other_in}')

Number of rows where CONTRIBUTING FACTOR VEHICLE 1 is missing, but at least one of CONTRIBUTING FACTOR VEHICLE 2-5 is provided: 0


In [74]:
# Length of dataset
df_len = df.shape[0]

missing_contributing_factor = df['CONTRIBUTING FACTOR VEHICLE 1'].isnull().sum()
missing_contributing_factor_percentage = (missing_contributing_factor / df_len) * 100


print(f'Number of rows where CONTRIBUTING FACTOR VEHICLE 1-5 is missing: {missing_contributing_factor}, which is {missing_contributing_factor_percentage:.2f}% of main dataset.')

Number of rows where CONTRIBUTING FACTOR VEHICLE 1-5 is missing: 3112, which is 0.27% of main dataset.


Checking how many unique values are for contributing factors and how often they apper in dataset

In [75]:
# Gathering all factors
factor_cols = [
    'CONTRIBUTING FACTOR VEHICLE 1',
    'CONTRIBUTING FACTOR VEHICLE 2',
    'CONTRIBUTING FACTOR VEHICLE 3',
    'CONTRIBUTING FACTOR VEHICLE 4',
    'CONTRIBUTING FACTOR VEHICLE 5'
]

all_factors = df[factor_cols].stack().str.strip().str.title()

# Checking how many unique values are for contributing factors
unique_count = all_factors.nunique()

print(f"Number of unique contributing factors: {unique_count}")

Number of unique contributing factors: 59


In [76]:
factor_counts = all_factors.value_counts()
factor_counts.head(60)

Unspecified                                              1334476
Driver Inattention/Distraction                            274558
Failure To Yield Right-Of-Way                              77869
Following Too Closely                                      71841
Other Vehicular                                            57435
Backing Unsafely                                           49273
Fatigued/Drowsy                                            37724
Passing Or Lane Usage Improper                             36210
Turning Improperly                                         34500
Passing Too Closely                                        31647
Unsafe Lane Changing                                       27243
Traffic Control Disregarded                                21413
Driver Inexperience                                        20835
Lost Consciousness                                         16916
Prescription Medication                                    12821
Pavement Slippery        

The most frequently occurring value for the contributing factor is 'Unspecified'. To avoid losing any potentially important data, it was decided to overwrite all missing values in CONTRIBUTING FACTOR VEHICLE 1 with 'Unspecified', since at least one vehicle was involved in each collision.

In [77]:
df['CONTRIBUTING FACTOR VEHICLE 1'] = df['CONTRIBUTING FACTOR VEHICLE 1'].fillna("Unspecified")

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;1.2.5.2 Checking VEHICLE TYPE CODE 1-5

In [78]:
mask = (
    df['VEHICLE TYPE CODE 1'].isnull() & (
    df['VEHICLE TYPE CODE 2'].notnull() |
    df['VEHICLE TYPE CODE 3'].notnull() |
    df['VEHICLE TYPE CODE 4'].notnull() |
    df['VEHICLE TYPE CODE 5'].notnull())
)

first_vehicle_missing_other_in = df[mask].shape[0]
print(f'Number of rows where VEHICLE TYPE CODE 1 is missing, but at least one of VEHICLE TYPE CODE 2-5 is provided: {first_vehicle_missing_other_in}')

Number of rows where VEHICLE TYPE CODE 1 is missing, but at least one of VEHICLE TYPE CODE 2-5 is provided: 4


In [79]:
df[mask].head()

DATETIME    BOROUGH  \
1041076 2015-05-05 11:40:00  MANHATTAN   
1041543 2015-04-14 13:00:00  MANHATTAN   
1158522 2015-03-30 12:45:00  MANHATTAN   
1251689 2014-12-29 19:00:00  MANHATTAN   

                                                  LOCATION  \
1041076  {'type': 'Point', 'coordinates': [-74.0072736,...   
1041543  {'type': 'Point', 'coordinates': [-74.0102592,...   
1158522  {'type': 'Point', 'coordinates': [-74.0157846,...   
1251689  {'type': 'Point', 'coordinates': [-74.0025032,...   

                                   STREET  NUMBER OF PERSONS INJURED  \
1041076  HUDSON STREET                                             0   
1041543  CHAMBERS STREET                                           0   
1158522  SOUTH END AVENUE                                          0   
1251689  PEARL STREET                                              0   

         NUMBER OF PERSONS KILLED  NUMBER OF PEDESTRIANS INJURED  \
1041076                         0                              0   
1041543                         0                              0   
1158522                         0                              0   
1251689                         0                              0   

         NUMBER OF PEDESTRIANS KILLED  NUMBER OF CYCLIST INJURED  \
1041076                             0                          0   
1041543                             0                          0   
1158522                             0                          0   
1251689                             0                          0   

         NUMBER OF CYCLIST KILLED  NUMBER OF MOTORIST INJURED  \
1041076                         0                           0   
1041543                         0                           0   
1158522                         0                           0   
1251689                         0                           0   

         NUMBER OF MOTORIST KILLED CONTRIBUTING FACTOR VEHICLE 1  \
1041076                          0                   Unspecified   
1041543                          0                   Unspecified   
1158522                          0               Other Vehicular   
1251689                          0                   Unspecified   

        CONTRIBUTING FACTOR VEHICLE 2 CONTRIBUTING FACTOR VEHICLE 3  \
1041076                   Unspecified                           NaN   
1041543                   Unspecified                           NaN   
1158522       Prescription Medication                           NaN   
1251689                           NaN                           NaN   

        CONTRIBUTING FACTOR VEHICLE 4 CONTRIBUTING FACTOR VEHICLE 5  \
1041076                           NaN                           NaN   
1041543                           NaN                           NaN   
1158522                           NaN                           NaN   
1251689                           NaN                           NaN   

         COLLISION_ID VEHICLE TYPE CODE 1 VEHICLE TYPE CODE 2  \
1041076       3214569                 NaN   PASSENGER VEHICLE   
1041543       3202121                 NaN                TAXI   
1158522       3194318                 NaN             UNKNOWN   
1251689       3144358                 NaN             UNKNOWN   

        VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4 VEHICLE TYPE CODE 5  
1041076                 NaN                 NaN                 NaN  
1041543                 NaN                 NaN                 NaN  
1158522                 NaN                 NaN                 NaN  
1251689                 NaN                 NaN                 NaN

In [80]:
# Length of dataset
df_len = df.shape[0]

missing_vehicle_type = df['VEHICLE TYPE CODE 1'].isnull().sum()
missing_vehicle_type_percentage = (missing_vehicle_type / df_len) * 100


print(f'Number of rows where VEHICLE TYPE CODE 1-5 is missing: {missing_vehicle_type}, which is {missing_vehicle_type_percentage:.2f}% of main dataset.')

Number of rows where VEHICLE TYPE CODE 1-5 is missing: 4170, which is 0.36% of main dataset.


In [81]:
vehicle_type_cols = [
    'VEHICLE TYPE CODE 1',
    'VEHICLE TYPE CODE 2',
    'VEHICLE TYPE CODE 3',
    'VEHICLE TYPE CODE 4',
    'VEHICLE TYPE CODE 5'
]

# Define replacements for known inconsistent labels
vehicle_type_replacements = {
    'SPORT UTILITY / STATION WAGON': 'SPORT UTILITY / STATION WAGON',
    'SPORT UTILITY/STATION WAGON': 'SPORT UTILITY / STATION WAGON',
    'STATION WAGON / SPORT UTILITY': 'SPORT UTILITY / STATION WAGON',
    'STATION WAGON/SPORT UTILITY': 'SPORT UTILITY / STATION WAGON',
    'SPORT UTILITY VEHICLE / STATION WAGON': 'SPORT UTILITY / STATION WAGON',
    'STATION WAGON / SPORT UTILITY VEHICLE': 'SPORT UTILITY / STATION WAGON',
    'STATION WAGON/SPORT UTILITY VEHICLE': 'SPORT UTILITY / STATION WAGON',
    'SPORT UTILITY VEHICLE': 'SPORT UTILITY',
    'STATION WAGON': 'STATION WAGON',
    'BICYCLE': 'BIKE',
    'E-BIK': 'BIKE',
    '4 DR SEDAN': 'SEDAN',
    'MOTORBIKE': 'MOTORCYCLE',
    'MOTORSCOOTER': 'MOTORCYCLE',
    'SCOOTER': 'MOTORCYCLE',
    'MOPED': 'MOTORCYCLE',
    'E-SCO': 'MOTORCYCLE',
    'AMBUL': 'AMBULANCE',
    'TRACTOR TRUCK DIESEL': 'TRUCK',
    'TRACTOR TRUCK GASOLINE': 'TRUCK',
    'TANKER': 'TRUCK',
    'BEVERAGE TRUCK': 'TRUCK',
    'BU': 'BUS',
    'FIRE': 'FIRE TRUCK',
    'UNKNO': 'UNKNOWN',
    '3-DOOR': 'PASSENGER VEHICLE',
    'CONV': 'CONVERTIBLE',
    'REFRIGERATED VAN': 'VAN',
}

for col in vehicle_type_cols:
    df[col] = df[col].str.strip().str.upper().replace(vehicle_type_replacements)

all_vehicle_type = df[vehicle_type_cols].stack()
all_vehicle_type = all_vehicle_type.str.strip().str.upper()

all_vehicle_type = all_vehicle_type.replace(vehicle_type_replacements)

In [82]:
vehicle_types = all_vehicle_type.value_counts()
vehicle_types.iloc[0:60]

PASSENGER VEHICLE                 918292
SPORT UTILITY / STATION WAGON     616080
SEDAN                             229657
TAXI                               95136
UNKNOWN                            70221
PICK-UP TRUCK                      51419
VAN                                40198
OTHER                              33689
BIKE                               32877
BUS                                31214
SMALL COM VEH(4 TIRES)             20615
LARGE COM VEH(6 OR MORE TIRES)     18003
BOX TRUCK                          13603
MOTORCYCLE                         13567
LIVERY VEHICLE                     13311
TRUCK                               6931
TK                                  4278
AMBULANCE                           4197
CONVERTIBLE                         2328
DUMP                                1989
DS                                  1626
FLAT BED                            1283
PK                                  1264
FIRE TRUCK                          1198
VN              

#### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;1.2.5.3 Filling missing values in CONTRIBUTING FACTOR VEHICLE 1-5 and VEHICLE TYPE CODE 1-5

In [83]:
for i in range(1, 6):
    factor_col = f'CONTRIBUTING FACTOR VEHICLE {i}'
    vehicle_col = f'VEHICLE TYPE CODE {i}'
    
    mismatch = df[(df[factor_col].notna() & df[vehicle_col].isna()) | (df[factor_col].isna() & df[vehicle_col].notna())]

    mismatch_percentage = (mismatch.shape[0] / df_len) * 100
    
    print(f'Mismatches between {factor_col} and {vehicle_col}: {mismatch.shape[0]} rows make up {mismatch_percentage:.2f}% of main dataset.')

Mismatches between CONTRIBUTING FACTOR VEHICLE 1 and VEHICLE TYPE CODE 1: 4170 rows make up 0.36% of main dataset.
Mismatches between CONTRIBUTING FACTOR VEHICLE 2 and VEHICLE TYPE CODE 2: 82188 rows make up 7.14% of main dataset.
Mismatches between CONTRIBUTING FACTOR VEHICLE 3 and VEHICLE TYPE CODE 3: 24904 rows make up 2.16% of main dataset.
Mismatches between CONTRIBUTING FACTOR VEHICLE 4 and VEHICLE TYPE CODE 4: 19416 rows make up 1.69% of main dataset.
Mismatches between CONTRIBUTING FACTOR VEHICLE 5 and VEHICLE TYPE CODE 5: 3653 rows make up 0.32% of main dataset.


The most significant mismatch is seen for vehicle 2. It was decided to fill missing values in each corresponding column when the other column provides a value.

In [84]:
for i in range(1, 6):
    factor_col = f'CONTRIBUTING FACTOR VEHICLE {i}'
    vehicle_col = f'VEHICLE TYPE CODE {i}'

    # Count missing before filling
    missing_factor_before = df[factor_col].isna().sum()
    missing_vehicle_before = df[vehicle_col].isna().sum()

    # Fill missing factor where vehicle type is present
    df.loc[df[factor_col].isna() & df[vehicle_col].notna(), factor_col] = 'Unspecified'

    # Fill missing vehicle type where factor is present
    df.loc[df[vehicle_col].isna() & df[factor_col].notna(), vehicle_col] = 'Unknown'

    # Count missing after filling
    missing_factor_after = df[factor_col].isna().sum()
    missing_vehicle_after = df[vehicle_col].isna().sum()

    # Calculate how many were filled
    factor_filled = missing_factor_before - missing_factor_after
    vehicle_filled = missing_vehicle_before - missing_vehicle_after

    factor_filled_percentage = (factor_filled / df_len) * 100
    vehicle_filled_percentage = (vehicle_filled / df_len) * 100

    print(f'For vehicle {i}: Filled {factor_filled} ({factor_filled_percentage:.2f}%) missing contributing factors and {vehicle_filled} ({vehicle_filled_percentage:.2f}%) missing vehicle types.')

For vehicle 1: Filled 0 (0.00%) missing contributing factors and 4170 (0.36%) missing vehicle types.
For vehicle 2: Filled 21184 (1.84%) missing contributing factors and 61004 (5.30%) missing vehicle types.
For vehicle 3: Filled 21535 (1.87%) missing contributing factors and 3369 (0.29%) missing vehicle types.
For vehicle 4: Filled 18759 (1.63%) missing contributing factors and 657 (0.06%) missing vehicle types.
For vehicle 5: Filled 3491 (0.30%) missing contributing factors and 162 (0.01%) missing vehicle types.


In [85]:
df.isna().sum()

DATETIME                               0
BOROUGH                                0
LOCATION                           80552
STREET                             45613
NUMBER OF PERSONS INJURED              0
NUMBER OF PERSONS KILLED               0
NUMBER OF PEDESTRIANS INJURED          0
NUMBER OF PEDESTRIANS KILLED           0
NUMBER OF CYCLIST INJURED              0
NUMBER OF CYCLIST KILLED               0
NUMBER OF MOTORIST INJURED             0
NUMBER OF MOTORIST KILLED              0
CONTRIBUTING FACTOR VEHICLE 1          0
CONTRIBUTING FACTOR VEHICLE 2     134479
CONTRIBUTING FACTOR VEHICLE 3    1055695
CONTRIBUTING FACTOR VEHICLE 4    1116811
CONTRIBUTING FACTOR VEHICLE 5    1143479
COLLISION_ID                           0
VEHICLE TYPE CODE 1                    0
VEHICLE TYPE CODE 2               134479
VEHICLE TYPE CODE 3              1055695
VEHICLE TYPE CODE 4              1116811
VEHICLE TYPE CODE 5              1143479
dtype: int64

### &nbsp;&nbsp;&nbsp;&nbsp;1.3 Exporting cleaned data to .csv file

In [86]:
df = df.reset_index(drop=True)

In [87]:
df.to_csv('cleaned_data.csv', index=False)

In [88]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1150944 entries, 0 to 1150943
Data columns (total 23 columns):
 #   Column                         Non-Null Count    Dtype         
---  ------                         --------------    -----         
 0   DATETIME                       1150944 non-null  datetime64[ns]
 1   BOROUGH                        1150944 non-null  category      
 2   LOCATION                       1070392 non-null  category      
 3   STREET                         1105331 non-null  object        
 4   NUMBER OF PERSONS INJURED      1150944 non-null  int8          
 5   NUMBER OF PERSONS KILLED       1150944 non-null  int8          
 6   NUMBER OF PEDESTRIANS INJURED  1150944 non-null  int8          
 7   NUMBER OF PEDESTRIANS KILLED   1150944 non-null  int8          
 8   NUMBER OF CYCLIST INJURED      1150944 non-null  int8          
 9   NUMBER OF CYCLIST KILLED       1150944 non-null  int8          
 10  NUMBER OF MOTORIST INJURED     1150944 non-null  int8 

## 2. Data exploration

In 02_Data_Exploration_DA-Project-II

## 3. Data visualization and insights

In 03_Data_Visualization_DA-Project-III